### PR TITLE
Can now pass urls to commandline fairtally with progressbar

### PR DESCRIPTION
**Pull request by jspaaks**
In this PR:

- Made `fairtally` accept urls via the commandline as arguments (#3)
- Added progressbar via `tqdm` (#23)
- Added capturing of stdout and stderr, store for later reporting
- Added pre-commit githook (#15) 
- Elevated rate limits already work (#9)
- Results are stored as array of object with keys (#5):
    ```

    {
       url: string,
       badge: string,
       repository: bool,
       license: bool,
       registry: bool,
       citation: bool,
       checklist: bool,
       count: int,
       stdout: string,
       stderr: string
    }
    ```
- Added `--html` and ``--json`` options. If specified, fairtally will write to a json file or to an html file, respectively. html writing uses jinja2 templating and `data/index.html.template`.

For testing, you can put in your own URLs, but for convenience, the snippet below gives you the 10 URLs from the Research Software Directory which were changed most recently:

```javascript
// Example to get a list of 10 most recently changed software items from Research Software Directory:
fetch("https://research-software.nl/api/software?limit=10&sort=updatedAt&direction=desc")
   .then((response) => {return response.json()})
   .then((data) => {
      console.log(data.map((x) => {
         return x.repositoryURLs.github[0]
      }))
   })
```

To test:

```
fairtally https://github.com/fair-software/howfairis https://gitlab.com/jspaaks/howfairis-livetest
```

Should show a progress bar and print results at the end

```
$ fairtally https://github.com/fair-software/howfairis https://gitlab.com/jspaaks/howfairis-livetest
fairtally progress: |██████████████████████                      | 1/2
currently checking https://gitlab.com/jspaaks/howfairis-livetest
```

If you have `jq` installed, you can pipe the results into it, like so:

```
fairtally <URL1> <URL2> <URL3> | jq '.'
```

or write to a json file:
```
fairtally --json results.json <URL1> <URL2> <URL3>
```

or write to an html file:
```
fairtally --html results.html <URL1> <URL2> <URL3>
```

Here's what that looks like in the browser:

image

### CODE REVIEW

File: fairtally/__version__.py [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
-import sys
+# import os
+# import sys
+#
+# # To update the package version number, edit CITATION.cff
+# citationfile = os.path.join(sys.exec_prefix, 'citation/fairtally', 'CITATION.cff')
+# with open(citationfile, 'r') as cff:
+#     for line in cff:
+#         if 'version:' in line:
+#             __version__ = line.replace('version:', '').strip().strip('"')
```
Comment (fdiblen) [inline fairtally/__version__.py:9]:
```suggestion
```

File: fairtally/cli.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+from howfairis import Repo
+from jinja2 import Template
+from tqdm import tqdm
+from fairtally.get_badge_color import get_badge_color
+from fairtally.redirect_stdout_stderr import RedirectStdStreams
+
+
+@click.command()
+@click.argument("urls", nargs=-1)
+@click.option("-o", "--output-filename", "output_file", help="Where to write the results of the analysis.",
```
Comment (fdiblen) [inline fairtally/cli.py:16]: Mention output formats in the help text

File: fairtally/cli.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        print(json.dumps(results))
+
+    elif output_file.name.endswith(".json"):
+        write_as_json()
+
+    elif output_file.name.endswith(".html"):
+        write_as_html()
+
+    else:
+        raise Exception("shouldnt happen")
```
Comment (fdiblen) [inline fairtally/cli.py:72]:
```suggestion
        raise Exception("Make sure the output file ends with '.html' or '.json'.")
```

File: fairtally/data/index.html.template [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  {% for item in results %}
+  <tr>
+    <td><a href="{{item.url}}">{{item.url}}</a></td>
+    <td>{{item.repository}}</td>
+    <td>{{item.license}}</td>
+    <td>{{item.registry}}</td>
+    <td>{{item.citation}}</td>
+    <td>{{item.checklist}}</td>
+    <td>{{item.count}}</td>
+    <td><a href="https://fair-software.nl"><img src="{{item.badge}}"></a></td>
```
Comment (fdiblen) [inline fairtally/data/index.html.template:30]:
```suggestion
    <td><a href="https://fair-software.eu"><img src="{{item.badge}}"></a></td>
```

File: setup.py [hunk 5]
```diff
@@ -1,62 +1,78 @@
...
-# To update the package version number, edit CITATION.cff
-with open('CITATION.cff', 'r') as cff:
-    for line in cff:
-        if 'version:' in line:
-            version = line.replace('version:', '').strip().strip('"')
+# # To update the package version number, edit CITATION.cff
+# with open("CITATION.cff", "r") as cff:
+#     for line in cff:
+#         if "version:" in line:
+#             version = line.replace("version:", "").strip().strip('"')
```
Comment (fdiblen) [inline setup.py:12]:
```suggestion
```

File: fairtally/cli.py (hunk 2 above)
Comment (jspaaks) [reply fairtally/cli.py]: I split this up into a `--html` and a `--json`, each with separate help text; this also avoids having to run the tool twice if you happen to want both json and html output.

**Comment by jspaaks** [comment]
The pep8 linting check fails but that's because those rules aren't helping me write clean code. E.g. the linelength of 79, how function calls break over multiple lines, the fact that you cannot have an empty line after a function signature.

File: setup.py [hunk 6]
```diff
@@ -1,62 +1,73 @@
...
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9"
+    ],
+    test_suite="tests",
+    install_requires=[
+        "click == 7.*",
+        "howfairis == 0.13.*",
+        "jinja2 == 2.*",
+        "requests == 2.*",
```
Comment (sverhoeven) [inline setup.py:46]: Requests is not used directly in code.
```suggestion
```

File: fairtally/cli.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    def write_as_html():
+        parent = Path(__file__).parent
+        template_file = parent / "data" / "index.html.template"
+        with open(template_file) as f:
+            template = Template(f.read())
+        s = template.render(results=results)
+        with output_file_html:
+            output_file_html.write(s)
+
+    if urls is None:
```
Comment (sverhoeven) [inline fairtally/cli.py:37]: Urls var is empty set when no urls are given.
```suggestion
    if len(urls) == 0:
```

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Tests commands and pre-commit script ran correctly. 
Spawned of some issues and left some inline comments for things that would otherwise never be done.

Great work.

File: fairtally/cli.py (hunk 7 above)
Comment (jspaaks) [reply fairtally/cli.py:37]: Indeed you're right

**Comment by jspaaks** [comment]
committed the suggestions

**Review by sverhoeven — APPROVED** [review]

**Review by eriktks — CHANGES_REQUESTED** [review]
Testing html output results in an error message:
```
$ pip install .
...
$ fairtally --html results.html https://github.com/fair-software/howfairis https://gitlab.com/jspaaks/howfairis-livetest
...
FileNotFoundError: [Errno 2] No such file or directory: 
'/home/erikt/anaconda3/envs/python37/lib/python3.7/site-packages/fairtally/data/index.html.template'
```

**Comment by jspaaks** [comment]
> Testing html output results in an error message:
> 
> ```
> $ pip install .
> ...
> $ fairtally --html results.html https://github.com/fair-software/howfairis https://gitlab.com/jspaaks/howfairis-livetest
> ...
> FileNotFoundError: [Errno 2] No such file or directory: 
> '/home/erikt/anaconda3/envs/python37/lib/python3.7/site-packages/fairtally/data/index.html.template'
> ```

True, but will be fixed by #25. See also https://github.com/fair-software/fairtally/issues/26. Do you want me to fix it here as well? My guess is we're eventually going to merge #25 anyway right?

**Review by eriktks — APPROVED** [review]
No, the missing file issue does not need to be fixed here if it is already fixed in another issue. Thanks!

**Review by fdiblen — APPROVED** [review]
Works nicely. Thanks @jspaaks

**Comment by sverhoeven** [comment]
Thanks for reviewing, merged on behalve of @jspaaks

**Pull request merged**